crates/sui-types/src/multisig.rs

```diff
@@ -0,0 +1,270 @@
+mod multisig_tests;
+
+pub type WeightUnit = u8;
+pub type ThresholdUnit = u16;
+pub const MAX_SIGNER_IN_MULTISIG: usize = 10;
+
+/// This initialize the underlying bytes representation of MultiSig. It encodes
+/// [struct MultiSig] as the MultiSig flag (0x03) concat with the bcs bytes
+/// of [struct MultiSig] i.e. `flag || bcs_bytes(MultiSig)`.
+impl AsRef<[u8]> for MultiSig {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes
+            .get_or_try_init::<_, eyre::Report>(|| {
+                let mut bytes = Vec::new();
+                bytes.push(SignatureScheme::MultiSig.flag());
+                bytes.extend_from_slice(
+                    bcs::to_bytes(self)
+                        .expect("BCS serialization should not fail")
+                        .as_slice(),
+                );
+                Ok(bytes)
+            })
+            .expect("OnceCell invariant violated")
+    }
+}
+
+/// The struct that contains signatures and public keys necessary for authenticating a MultiSig.
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+pub struct MultiSig {
+    /// The plain signature encoded with signature scheme.
+    sigs: Vec<CompressedSignature>,
+    /// A bitmap that indicates the position of which public key the signature should be authenticated with.
+    #[schemars(with = "Base64")]
+    #[serde_as(as = "SuiBitmap")]
+    bitmap: RoaringBitmap,
+    /// The public key encoded with each public key with its signature scheme used along with the corresponding weight.
+    multisig_pk: MultiSigPublicKey,
+    /// A bytes representation of [struct MultiSig]. This helps with implementing [trait AsRef<[u8]>].
+    #[serde(skip)]
+    bytes: OnceCell<Vec<u8>>,
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl PartialEq for MultiSig {
+    fn eq(&self, other: &Self) -> bool {
+        self.sigs == other.sigs
+            && self.bitmap == other.bitmap
+            && self.multisig_pk == other.multisig_pk
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Eq for MultiSig {}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Hash for MultiSig {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+impl AuthenticatorTrait for MultiSig {
+    fn verify_secure_generic<T>(
+        &self,
+        value: &IntentMessage<T>,
+        author: SuiAddress,
+    ) -> Result<(), SuiError>
+    where
+        T: Serialize,
+    {
+        if self.multisig_pk.pk_map.len() > MAX_SIGNER_IN_MULTISIG {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid number of public keys".to_string(),
+            });
+        }
+
+        if <SuiAddress as From<MultiSigPublicKey>>::from(self.multisig_pk.clone()) != author {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid address".to_string(),
+            });
+        }
+        let mut weight_sum = 0;
```

Comment:
I'd specifically add the type here (u16) just for readability. I understand we have weights being u8 atm and we aggregate them into a u16, but in any case I'd also add a test where all keys have max weight (just to capture any future update where for some reason we allow 256+ pubKeys)